File: main/java/Project/Tools/ImageDrawer.java
package Project.Tools;

import com.madgag.gif.fmsware.AnimatedGifEncoder;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.net.URL;

import static Project.Tools.JImageDrawerUtils.*;

public class ImageDrawer {

    /**
     * 推 的动画
     *
     * @param files   gif集
     * @param oFile   要被推的 url
     * @param outFile 输出
     * @return
     * @throws Exception
     */
    public static File getTuiGift(File[] files, URL oFile, File outFile) throws Exception {
        AnimatedGifEncoder encoder = new AnimatedGifEncoder();
        encoder.start(outFile.getAbsolutePath());
        encoder.setRepeat(0);
        encoder.setQuality(5);
        encoder.setFrameRate(200);
        int rotateEve = 360 / files.length;
        final BufferedImage oImage = ImageIO.read(oFile);
        for (int i = 0; i < files.length; i++) {
            encoder.setDelay(100);
            BufferedImage main = ImageIO.read(files[i]);
            float rotate = rotateEve * i;
            BufferedImage image = oImage;
            image = (BufferedImage) Image2Size(image, 200, 200);
            image = (BufferedImage) rotateImage(image, rotate);
            image = roundImage(image, 9999);
            image = putImage(main, image, 93, 83);
            encoder.addFrame(image);
        }
        encoder.finish();
        return outFile;
    }

    public static void main(String[] args) throws Exception {
//        System.out.println(getTuiGift(new File("./images/tui").listFiles()
//                , new URL("https://q1.qlogo.cn/g?b=qq&nk=3474006766&s=640"), new File("./data/end1.gif")));
    }

    public static File getWq(File[] files, URL oFile, File outFile) throws Exception {
//        File[] files = new File("./data/wq").listFiles();
//        File outFile = new File("./data/end.gif");
//        URL oFile = new URL("https://q1.qlogo.cn/g?b=qq&nk=3474006766&s=640");
        AnimatedGifEncoder encoder = new AnimatedGifEncoder();
        encoder.start(outFile.getAbsolutePath());
        encoder.setRepeat(0);
        encoder.setQuality(5);
        encoder.setFrameRate(200);
        final BufferedImage oImage = ImageIO.read(oFile);
        int rotateEve = 360 / files.length;
        for (int i = 0; i < files.length; i++) {
            encoder.setDelay(100);
            BufferedImage main = ImageIO.read(files[i]);
            BufferedImage image = oImage;
            float rotate = rotateEve * i;
            image = (BufferedImage) Image2Size(image, 132, 132);
            image = (BufferedImage) rotateImage(image, rotate);
            image = roundImage(image, 9999);
            int[] vs = getWt(i);
            image = putImage(main, image, vs[0], vs[1]);
            encoder.addFrame(image);
            ImageIO.write(image,"png",new File("./data/temp/"+i+".png"));
        }
        encoder.finish();
        return outFile;
    }

    private static int[] getWt(int i) {
        int x = 52, y = 27;
        switch (i) {
            case 2:
                y += 8;
                x -= 2;
                break;
            case 3:
                y += 6;
                x -= 2;
                break;
            case 4:
                y += 7;
                x -= 2;
                break;
            case 5:
                y += 7;
                x -= 3;
                break;
            case 6:
                y += 3;
                x -= 3;
                break;
            case 7:
                y += 1;
                x -= 3;
                break;
            case 8:
            case 9:
                x -= 3;
                break;
            case 10:
                y -= 5;
                x -= 3;
                break;
            case 11:
                y -= 5;
                x -= 5;
                break;
            case 13:
                y -= 4;
                x -= 4;
                break;
            case 14:
                y += 3;
                x -= 2;
                break;
            case 15:
                y += 2;
                x -= 2;
                break;
            case 16:
                x -= 1;
                y += 1;
                break;
            case 25:
                y += 3;
                x += 3;
                break;
            case 26:
                y += 4;
                break;
            case 27:
                y += 4;
                break;
            case 28:
            case 29:
                x -= 1;
                y += 3;
                break;

        }
        return new int[]{x, y};
    }
}

File: main/java/Project/Tools/ImageDrawer.kt
package Project.Tools

import kotlin.Throws
import com.madgag.gif.fmsware.AnimatedGifEncoder
import java.awt.image.BufferedImage
import javax.imageio.ImageIO
import Project.Tools.ImageDrawer
import java.awt.geom.RoundRectangle2D
import java.io.IOException
import net.coobird.thumbnailator.Thumbnails
import net.coobird.thumbnailator.geometry.Positions
import java.awt.*
import java.io.File
import java.lang.Exception
import java.net.URL

object ImageDrawer {
    /**
     * 推 的动画
     *
     * @param files   gif集
     * @param oFile   要被推的 url
     * @param outFile 输出
     * @return
     * @throws Exception
     */
    @JvmStatic
    @Throws(Exception::class)
    fun getTuiGift(files: Array<File?>, oFile: URL?, outFile: File): File {
        val encoder = AnimatedGifEncoder()
        encoder.start(outFile.absolutePath)
        encoder.setRepeat(0)
        encoder.setQuality(5)
        encoder.setFrameRate(200f)
        val rotateEve = 360 / files.size
        for (i in files.indices) {
            encoder.setDelay(100)
            val main = ImageIO.read(files[i])
            var image = ImageIO.read(oFile)
            val rotate = (rotateEve * i).toFloat()
            image = Image2Size(image, 200, 200) as BufferedImage
            image = rotateImage(image, rotate) as BufferedImage
            image = roundImage(image, 9999)
            image = putImage(main, image, 93, 83)
            encoder.addFrame(image)
        }
        encoder.finish()
        return outFile
    }

    /**
     * 图片圆角
     *
     * @param image        图片
     * @param cornerRadius 幅度
     * @return
     */
    fun roundImage(image: BufferedImage, cornerRadius: Int): BufferedImage {
        val width = image.width
        val height = image.height
        val outputImage = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
        val g2 = outputImage.createGraphics()
        g2.composite = AlphaComposite.Src
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
        g2.color = Color.WHITE
        g2.fill(
            RoundRectangle2D.Float(
                0f, 0f,
                width.toFloat(),
                height.toFloat(),
                cornerRadius.toFloat(),
                cornerRadius.toFloat()
            )
        )
        g2.composite = AlphaComposite.SrcAtop
        g2.drawImage(image, 0, 0, null)
        g2.dispose()
        return outputImage
    }

    /**
     * 压缩指定宽、高
     *
     * @param bimg
     * @param width
     * @param height
     * @param tagFilePath
     * @return
     */
    @Throws(IOException::class)
    fun Image2Size(bimg: BufferedImage?, width: Int, height: Int): Image {
        var tempFile: File? = null
        try {
            tempFile = File.createTempFile("temp1", ".png")
            Thumbnails.of(bimg).size(width, height).outputQuality(1f).toFile(tempFile)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        val image = ImageIO.read(tempFile)
        tempFile!!.delete()
        return image
    }

    /**
     * 旋转图片
     *
     * @param image
     * @param rotate
     * @return
     * @throws IOException
     */
    @Throws(IOException::class)
    fun rotateImage(image: BufferedImage, rotate: Float): Image {
        var image = image
        var tempFile: File? = null
        try {
            val _w = image.width
            val _h = image.height
            tempFile = File.createTempFile("temp2", ".png")
            Thumbnails.of(image).scale(1.0).rotate(rotate.toDouble()).toFile(tempFile)
            val i1 = ImageIO.read(tempFile)
            val w = i1.width
            val h = i1.height
            Thumbnails.of(i1)
                .sourceRegion(Positions.CENTER, 200, 200)
                .size(200, 200)
                .keepAspectRatio(false)
                .toFile(tempFile)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        image = ImageIO.read(tempFile)
        tempFile!!.delete()
        return image
    }

    /**
     * 将 一张图片放到 一张图片上
     *
     * @param image
     * @param im
     * @param x
     * @param y
     * @return
     */
    fun putImage(image: BufferedImage, im: BufferedImage?, x: Int, y: Int): BufferedImage {
        val graphics = image.graphics
        graphics.drawImage(im, x, y, null)
        return image
    }
}